src/Preprocessor.zig

```diff
@@ -510,14 +513,400 @@ fn skipToNl(tokenizer: *Tokenizer) void {
+                };
+
+                var res = try pp.pasteTokens2(prev, next);
+                defer pp.comp.gpa.free(res);
+                // TODO: more work than necessary, but easy
+                try buf.replaceRange(buf.items.len - prev.len, prev.len, res);
+                // skip next token
+                tok_i += 1;
+            },
+            .macro_param => {
+                // TODO: raw.end contains the index of the param
+                const arg = expanded_args.items[raw.end];
+
+                if (arg.len == 0) {
+                    // needed for the following token pasting phase
+                    try buf.append(.{
+                        .id = .empty_arg,
+                        .loc = .{ .id = raw.source, .byte_offset = raw.start }
+                    });
+                } else {
+                    for (arg) |tok| {
+                        try buf.ensureCapacity(buf.items.len + arg.len);
+                        buf.appendAssumeCapacity(tok);
+                    }
+                }
+            },
+            .keyword_va_args => {
+                try buf.ensureCapacity(buf.items.len + expanded_variable_arguments.items.len);
+                buf.appendSliceAssumeCapacity(expanded_variable_arguments.items);
+            },
+            .stringify_param, .stringify_va_args => {
+                const arg = if (raw.id == .stringify_va_args)
+                    variable_arguments.items
+                else
+                    args.items[raw.end];
+
+                var char_buf = std.ArrayList(u8).init(pp.comp.gpa);
+                defer char_buf.deinit();
+
+                // TODO pretty print these
+                try char_buf.append('"');
+                for (arg) |tok, i| {
+                    if (i != 0) try char_buf.append(' ');
+                    for (pp.expandedSlice(tok)) |c| {
+                        if (c == '"')
+                            try char_buf.appendSlice("\\\"")
+                        else
+                            try char_buf.append(c);
+                    }
+                }
+                try char_buf.appendSlice("\"\n");
+
+                const start = pp.generated.items.len;
+                try pp.generated.appendSlice(char_buf.items);
+
+                try buf.append(.{
+                    .id = .string_literal,
+                    .loc = .{ // location of token slice in the generated buffer
+                        .id = .generated,
+                        .byte_offset = @intCast(u32, start),
+                    },
+                });
+            },
+            else => {
+                try buf.append(tokFromRaw(raw));
+            }
+        }
+    }
+
+    return buf;
+}
+
+fn shouldExpand(tok: Token, macro: *Macro) bool {
+    const macro_loc = switch (macro.*) {
+        .simple => |smacro| smacro.loc,
+        .func => |smacro| smacro.loc,
+        else => unreachable
+    };
+    var maybe_loc = tok.loc.next;
+    while (maybe_loc) |loc| {
```

Comment:
A future improvement will be to convert the source location mechanism into using contiguous arrays instead of linked lists now that they are being iterated upon a lot more.

Reply:
Yeah, I will open a issue/pr after this one is done